### ISSUE TITLE
Why does batch size affect convergence?

### ISSUE THREAD

**Issue by 0dB** [karpathy/nanoGPT#285]
It is my understanding from the videos that batch size should have no influence on convergence. But I have cases where increasing the batch size will lead to underfitting, or where decreasing the batch size will lead to overfitting. Is there an explanation for this?  Am I getting something wrong? (AFAIK batch normalization could have this effect, but not layer normalization.)

**Comment by matthiasgeihs** [comment]
I'm not necessarily an expert but I have some intuition why this might happen:
Let's say you have a very small batch size. This means there is only very few samples in your batch. Now you basically start optimizing per batch. But since there is only little entropy in each batch, your training procedure will not converge nicely. Does that make sense?

PS: This image somewhat summarizes it:
![image](https://github.com/karpathy/nanoGPT/assets/62935430/7f589b03-3a50-42be-aed7-fa6c23b420cc)

**Comment by 0dB** [comment]
Hm, from looking at the code, I think that in nanoGPT `batch_size` is actually the number of batches and not the _size_ of the batches, instead, `block_size` is the size of the batches (how much data is in a batch). If I am right, the explanation you give would make sense for the `block_size`. This is a bit confusing, I think. Do you read the function `get_batch()` in `train.py` like that too?

**Comment by matthiasgeihs** [comment]
You might want to watch https://youtu.be/kCc8FmEb1nY?t=867. This might clarify a few things.

**Comment by 0dB** [comment]
Right, that is the part of the videos I am referring to in my original question. Andrej says the blocks (batches) are "completely independent" and "don't talk to each other", but this is not the experience I have and I am wondering what is going on…

**Comment by matthiasgeihs** [comment]
*The different batches don't talk to each other* means that the  model parameters are optimized **per batch**.
```
block_size ^= context_length ^= length of a training chunk
batch_size ^= number of blocks (a.k.a. chunks) per training batch
```

**Revisiting the example in the video:** If we have `block_size = 8` and `batch_size = 4`, this actually makes up `8` training samples per block and `32 = 4 * 8` training samples per batch. The reason for this is that the transformer even without batching will process `block_size` samples in parallel (i.e., all the substrings `block[0:i]` for `i in range(1,9)`, with targets `block[i]`).

(Note that there is also the more advanced method of micro-batching (a.k.a. averaging over multiple batches) if you are running on multiple GPUs, but let's not look at that for now for the sake of simplicity.)

**Comment by 0dB** [comment]
Let me get back to you on that after I did some more code checking. Just one quick remark on what you wrote, and one more quote from Andrej:

> The reason for this is that the transformer even without batching will process block_size samples in parallel (i.e., all the substrings block[0:i] for i in range(1,9), with targets block[i]).

Righ, but what you are mentioning here is the T dimension (context size, i. e. size of the "window"), but I think the answer lies somewhere how the B dimension (number of blocks per batch) is handled. And there is already one clear case where the batches are not independent: As Andrej explains in the video and as can be seen in the function `estimate_loss()`, the loss is averaged over all batches. So that is one thing where the blocks are connected after all, but I don't know yet if that can explain the effect I am seeing.

And, another tangent on my original question: Andrej says in the video that the batches (blocks) are only so that the GPU is kept busy, so to me that would imply that convergence should not be dependent on whether I have 32 blocks run at once in one batch or one block per batch, so in this example 32 blocks run in 32 batches. Would you say so too? But I indeed see a difference regarding convergence when I change `batch_size` (i. e. number of blocks per batch). (I ran a few dozen tests over the course of a few weeks, all running for hours, on a 5MB Shakespeare data set, not the 1MB data set Andrej used, and while tuning the hyperparamers I noticed this effect and zeroed in on it.) I will study the code some more.

**Comment by matthiasgeihs** [comment]
I think you might wanna look at how which point the backprop optimization actually changes the parameters. This is done after each batch.

(Also note that batch != block.)

**Comment by matthiasgeihs** [comment]
wow, in this case i am also pretty much out of explanations. of course you can try to run with different seed. or maybe batch size has to be a multiple of something? idk

**Comment by chirkofun** [comment]
@0dB is there any explanation that you found for batch size affecting convergence? For me in general it is not clear what batch size changes except faster computation on GPU (MPS in my case)

**Comment by pyotr777** [comment]
@matthiasgeihs was correct here (https://github.com/karpathy/nanoGPT/issues/285#issuecomment-1592530587). It is the batch size that controls how many training tokens the model sees before updating the weights. 

block size, as in the comment:

`block_size: int = 1024  # max sequence length`

is the sequence length (also called 'context size').

batch size = block size * micro batch size * accumulation steps * number of parallel processes
